# A simple scatter plot
In this example, you're going to make a scatter plot of female literacy vs fertility using data from the European Environmental Agency. This dataset highlights that countries with low female literacy have high birthrates. The x-axis data has been loaded for you as fertility and the y-axis data has been loaded as female_literacy.

Your job is to create a figure, assign x-axis and y-axis labels, and plot female_literacy vs fertility using the circle glyph.

After you have created the figure, in this exercise and the ones to follow, play around with it! Explore the different options available to you on the tab to the right, such as "Pan", "Box Zoom", and "Wheel Zoom". You can click on the question mark sign for more details on any of these tools.

Note: You may have to scroll down to view the lower portion of the figure.

* Import the figure function from bokeh.plotting, and the output_file and show functions from bokeh.io.
* Create the figure p with figure(). It has two parameters: x_axis_label and y_axis_label.
* Add a circle glyph to the figure p using the function p.circle() where the inputs are, in order, the x-axis data and y-axis data.
* Use the output_file() function to specify the name 'fert_lit.html' for the output file.
* Create and display the output file using show() and passing in the figure p.

In [33]:
# Import figure from bokeh.plotting
from bokeh.plotting import figure

# Import output_file and show from bokeh.io
from bokeh.io import show, output_file

import os

if os.path.exists("demofile.txt"):
  os.remove("demofile.txt")

import pandas as pd
import numpy as np

df = pd.read_csv('literacy_birth_rate.csv')
df = df.dropna(how='any')

fertility = df['fertility'].astype('float64').values
female_literacy = df['female literacy'].astype('float64').values


# Create the figure: p
p = figure(x_axis_label='fertility (children per woman)', y_axis_label='female_literacy (% population)')

# Add a circle glyph to the figure p
p.circle(fertility, female_literacy)

# Call the output_file() function and specify the name of the file
output_file('fert_lit.html')

# Display the plot
show(p)


In [56]:
df

Country  Continent female literacy fertility  \
0                                Chine       ASI            90.5     1.769   
1                                 Inde       ASI            50.8     2.682   
2                                  USA       NAM              99     2.077   
3                            Indonésie       ASI            88.8     2.132   
4                               Brésil       LAT            90.2     1.827   
5                             Pakistan       ASI              40     3.872   
6                           Bangladesh       ASI            49.8     2.288   
7                              Nigéria        AF            48.8     5.173   
8                 Fédération de Russie       EUR            99.4     1.393   
9                                Japan       ASI              99     1.262   
10                             Mexique       LAT            91.5     2.156   
11                         Philippines       ASI            93.9     3.026   
12                            Viet Nam       ASI            90.2     2.033   
13                             Germany       EUR              99     1.324   
14                              Egypte        AF            57.8     2.816   
15                            Ethiopie        AF            22.8     5.211   
16                             Turquie       ASI            81.3       2.1   
17       Iran, République islamique d'       ASI            77.2     1.781   
18                           Thaïlande       ASI            91.5     1.822   
19          Rép. Démocratique du Congo        AF            56.1     5.908   
20                              France       EUR              99     1.881   
21                                  UK       EUR              99     1.852   
22                              Italie       EUR            98.5      1.39   
23                             Myanmar       ASI            89.2     2.281   
24                      Afrique du Sud        AF            88.1     2.505   
25                         South Korea       ASI            96.6     1.224   
26                             Ukraine       EUR            99.6     1.361   
27                             Espagne       EUR            96.9     1.468   
28                            Colombie       LAT            93.4     2.404   
29         République-Unie de Tanzanie        AF            66.3      5.52   
..                                 ...       ...             ...       ...   
132                            Lesotho        AF            95.1     3.264   
133  L'ex-Rép. yougoslave de Macédoine       EUR            95.4     1.436   
134                           Slovénie       EUR            99.7     1.393   
135                           Botswana        AF            83.5     2.822   
136                             Gambie        AF            34.3     4.969   
137                      Guinée-Bissau        AF            36.5     5.659   
138                              Gabon        AF            83.2      3.24   
139                            Estonie       EUR            99.8     1.693   
140                  Trinité-et-Tobago       LAT            98.2     1.647   
141                              Qatar       ASI            90.4      2.36   
142                            Maurice        AF            84.8     1.792   
143                          Swaziland        AF            85.6      3.45   
144                             Chypre       EUR            96.7     1.516   
145                            Bahreïn       ASI            89.4     2.233   
146                            Bhoutan       ASI            38.7     2.563   
147                 Guinée équatoriale        AF            89.1     5.283   
148                            Comores        AF            67.8     3.885   
149                       Macao, Chine       ASI            90.7     0.966   
150                           Suriname       LAT            88.4     2.373   
151                           Cap-Vert        AF    

# A scatter plot with different shapes
By calling multiple glyph functions on the same figure object, we can overlay multiple data sets in the same figure.

In this exercise, you will plot female literacy vs fertility for two different regions, Africa and Latin America. Each set of x and y data has been loaded separately for you as fertility_africa, female_literacy_africa, fertility_latinamerica, and female_literacy_latinamerica.

Your job is to plot the Latin America data with the circle() glyph, and the Africa data with the x() glyph.

figure has already been imported for you from bokeh.plotting.

* Create the figure p with the figure() function. It has two parameters: x_axis_label and y_axis_label.
* Add a circle glyph to the figure p using the function p.circle() where the inputs are the x and y data from Latin America: fertility_latinamerica and female_literacy_latinamerica.
* Add an x glyph to the figure p using the function p.x() where the inputs are the x and y data from Africa: fertility_africa and female_literacy_africa.
* The code to create, display, and specify the name of the output file has been written for you, so after adding the x glyph, hit 'Submit Answer' to view the figure.

In [64]:
fertility_latinamerica_df = df[df['Continent']=='LAT']
fertility_latinamerica = fertility_latinamerica_df['fertility'].astype('float64').values
female_literacy_latinamerica = fertility_latinamerica_df['female literacy'].astype('float64').values

fertility_africa_df = df[df['Continent']=='AF']
fertility_africa = fertility_africa_df['fertility'].astype('float64').values
female_literacy_africa = fertility_africa_df['female literacy'].astype('float64').values

# Create the figure: p
p = figure(x_axis_label='fertility', y_axis_label='female_literacy (% population)')

# Add a circle glyph to the figure p
p.circle(fertility_latinamerica, female_literacy_latinamerica)

# Add an x glyph to the figure p
p.x(fertility_africa,female_literacy_africa)

# Specify the name of the file
output_file('fert_lit_separate.html')

# Display the plot
show(p)

# Customizing your scatter plots
The three most important arguments to customize scatter glyphs are color, size, and alpha. Bokeh accepts colors as hexadecimal strings, tuples of RGB values between 0 and 255, and any of the 147 CSS color names. Size values are supplied in screen space units with 100 meaning the size of the entire figure.

The alpha parameter controls transparency. It takes in floating point numbers between 0.0, meaning completely transparent, and 1.0, meaning completely opaque.

In this exercise, you'll plot female literacy vs fertility for Africa and Latin America as red and blue circle glyphs, respectively.

* Using the Latin America data (fertility_latinamerica and female_literacy_latinamerica), add a blue circle glyph of size=10 and alpha=0.8 to the figure p. To do this, you will need to specify the color, size and alpha keyword arguments inside p.circle().
* Using the Africa data (fertility_africa and female_literacy_africa), add a red circle glyph of size=10 and alpha=0.8 to the figure p.

In [66]:
# Create the figure: p
p = figure(x_axis_label='fertility (children per woman)', y_axis_label='female_literacy (% population)')

# Add a blue circle glyph to the figure p
p.circle(fertility_latinamerica, female_literacy_latinamerica, color = 'blue', size = 10, alpha = 0.8)

# Add a red circle glyph to the figure p
p.circle(fertility_africa, female_literacy_africa, color='red', size=10, alpha=0.8)

# Specify the name of the file
output_file('fert_lit_separate_colors.html')

# Display the plot
show(p)

# Lines
We can draw lines on Bokeh plots with the line() glyph function.

In this exercise, you'll plot the daily adjusted closing price of Apple Inc.'s stock (AAPL) from 2000 to 2013.

The data points are provided for you as lists. date is a list of datetime objects to plot on the x-axis and price is a list of prices to plot on the y-axis.

Since we are plotting dates on the x-axis, you must add x_axis_type='datetime' when creating the figure object.

* Import the figure function from bokeh.plotting.
* Create a figure p using the figure() function with x_axis_type set to 'datetime'. The other two parameters are x_axis_label and y_axis_label.
* Plot date and price along the x- and y-axes using p.line().

In [69]:
df = pd.read_csv('aapl.csv')
df.head()

date = pd.to_datetime(df['date'])
price = df['adj_close']

# Import figure from bokeh.plotting
from bokeh.plotting import figure

# Create a figure with x_axis_type="datetime": p
p = figure(x_axis_type = 'datetime', x_axis_label='Date', y_axis_label='US Dollars')

# Plot date along the x axis and price along the y axis
p.line(date, price)
# Specify the name of the output file and show the result
output_file('line.html')
show(p)
